test/gurka/test_time_tracking.cc

```diff
@@ -174,86 +174,82 @@ TEST(TimeTracking, reverse) {
-      {"EF", {{"highway", "primary"}}},  {"FG", {{"highway", "primary"}}},
-      {"GH", {{"highway", "primary"}}},
+      {"AB", {{"highway", "residential"}}}, {"BC", {{"highway", "residential"}}},
+      {"CD", {{"highway", "residential"}}}, {"DE", {{"highway", "residential"}}},
+      {"EF", {{"highway", "primary"}}},     {"FG", {{"highway", "primary"}}},
+      {"GH", {{"highway", "primary"}}},     {"DI", {{"highway", "residential"}}},
+      {"IJ", {{"highway", "primary"}}},     {"JE", {{"highway", "primary"}}},
   };
   const auto layout = gurka::detail::map_to_coordinates(ascii_map, 100);
   auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_time_tracking_make",
                                {{"mjlonir.timezone", "/path/to/timezone.sqlite"}});
 
-  // pick out a start and end ll by finding the appropriate edges in the graph
-  baldr::GraphReader reader(map.config.get_child("mjolnir"));
-  auto start = map.nodes["A"];
-  auto end = map.nodes["H"];
-
-  // route between them with a current time
-  auto req =
-      boost::format(
-          R"({"costing":"auto","date_time":{"type":0},"locations":[{"lon":%1%,"lat":%2%},{"lon":%3%,"lat":%4%}]})") %
-      start.first % start.second % end.first % end.second;
-  valhalla::Api api;
-  tyr::actor_t actor(map.config, reader);
-  actor.route(req.str(), nullptr, &api);
-
-  // check the timings
+  const std::vector<double> expected = {0,       51.4286, 102.857, 154.286, 195.446, 0,
+                                        148.927, 169.104, 215.799, 239.817, 263.817, 287.817};
+
+  auto test_result = [&expected](const std::vector<double>& actual_times, const Api& request) {
+    ASSERT_THAT(actual_times, testing::Pointwise(testing::DoubleNear(0.001), expected));
+    // first location has a waiting_time, second not since it was "through"
+    EXPECT_EQ(request.options().locations(1).waiting_secs(), 600.f);
+    EXPECT_EQ(request.options().locations(2).waiting_secs(), 0.f);
+    EXPECT_EQ(request.info().warnings(0).code(), 203);
+  };
+
   std::vector<double> times;
-  for (const auto& route : api.trip().routes()) {
+  auto result = gurka::do_action(Options::route, map, {"A", "I", "1", "H"}, "auto",
+                                 {{"/date_time/type", "0"},
+                                  {"/locations/1/waiting", "600"},
+                                  {"/locations/2/type", "through"},
+                                  {"/locations/2/waiting", "600"}});
+  for (const auto& route : result.trip().routes()) {
     for (const auto& leg : route.legs()) {
       for (const auto& node : leg.node()) {
         times.push_back(node.cost().elapsed_cost().seconds());
       }
     }
   }
-  const std::vector<double> expected = {0,        17.1429,  34.2857,  51.4286,
-                                        193.9319, 245.4273, 269.4273, 293.4273};
-  ASSERT_THAT(times, testing::Pointwise(testing::DoubleNear(0.0001), expected));
+  test_result(times, result);
```

Comment:
I tested all things here:
- that it's outputting the right `waiting` time
- it's ignoring `through` and
- issues a warning

Only JSON output is not tested I guess.